gobblin-core/src/main/java/gobblin/publisher/BaseDataPublisher.java

```diff
@@ -481,45 +507,44 @@ public void publishMetadata(WorkUnitState state)
 
     // For each branch - retrieve metadata and metadata output directory from task state; then publish
     for (int branchId = 0; branchId < this.numBranches; branchId++) {
-      String metadataValue = getMergedMetadataForState(state, branchId);
+      String metadataValue = getMergedMetadataForState(state, null, branchId);
       if (metadataValue == null) {
         //Nothing to write
         continue;
       }
       Path metadataOutputPath = getMetadataOutputFileForBranch(state, branchId);
-      publishMetadata(metadataValue, branchId, Collections.singleton(metadataOutputPath));
+      publishMetadata(metadataValue, branchId, metadataOutputPath);
     }
   }
 
   /**
    * Publish metadata to a set of paths
    */
-  private void publishMetadata(String metadataValue, int branchId, Collection<Path> pathsToPublish)
+
+  private void publishMetadata(String metadataValue, int branchId, Path metadataOutputPath)
       throws IOException {
-    for (Path metadataOutputPath : pathsToPublish) {
-      try {
-        if (metadataOutputPath == null) {
-          LOG.info("Metadata output path not set for branch " + String.valueOf(branchId) + ", not publishing.");
-          continue;
-        }
+    try {
+      if (metadataOutputPath == null) {
+        LOG.info("Metadata output path not set for branch " + String.valueOf(branchId) + ", not publishing.");
+        return;
+      }
 
-        FileSystem fs = this.metaDataWriterFileSystemByBranches.get(branchId);
+      FileSystem fs = this.metaDataWriterFileSystemByBranches.get(branchId);
 
-        if (!fs.exists(metadataOutputPath.getParent())) {
-          WriterUtils.mkdirsWithRecursivePermission(fs, metadataOutputPath, this.permissions.get(branchId));
-        }
+      if (!fs.exists(metadataOutputPath.getParent())) {
+        WriterUtils.mkdirsWithRecursivePermission(fs, metadataOutputPath, this.permissions.get(branchId));
+      }
 
-        //Delete the file if metadata already exists
-        if (fs.exists(metadataOutputPath)) {
-          HadoopUtils.deletePath(fs, metadataOutputPath, false);
-        }
-        LOG.info("Writing metadata for branch " + String.valueOf(branchId) + " to " + metadataOutputPath.toString());
-        try (FSDataOutputStream outputStream = fs.create(metadataOutputPath)) {
-          outputStream.write(metadataValue.getBytes(StandardCharsets.UTF_8));
-        }
-      } catch (IOException e) {
-        LOG.error("Metadata file is not generated: " + e, e);
+      //Delete the file if metadata already exists
+      if (fs.exists(metadataOutputPath)) {
```

Comment:
Should the metadata be merged instead of deleted? We have cases where an output directory can have additional data files added to an existing directory by a different job.